File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int m;
	static int n;
	static int[][] field;
	static boolean[][] visited;
	static int[] dx = {-1, 0, 1, 0};
	static int[] dy = {0, 1, 0, -1};

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine());

		for (int i = 0; i < t; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine(), " ");
			m = Integer.parseInt(st.nextToken());
			n = Integer.parseInt(st.nextToken());
			int k = Integer.parseInt(st.nextToken());

			field = new int[m][n];
			visited = new boolean[m][n];

			for (int j = 0; j < k; j++) {
				st = new StringTokenizer(br.readLine(), " ");
				int x = Integer.parseInt(st.nextToken());
				int y = Integer.parseInt(st.nextToken());
				field[x][y] = 1;
			}

			int count = 0;

			for (int a = 0; a < m; a++) {
				for (int b = 0; b < n; b++) {
					if (field[a][b] == 1 && !visited[a][b]) {
						dfs(a, b);
						count++;
					}
				}
			}
			System.out.println(count);
		}
	}

	private static void dfs(int x, int y) {
		visited[x][y] = true;

		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];

			if (nx >= 0 && nx < m && ny >= 0 && ny < n) {
				if (field[nx][ny] == 1 && !visited[nx][ny]) {
					dfs(nx, ny);
				}
			}
		}
	}
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys

sys.setrecursionlimit(10**6)

t = int(sys.stdin.readline().rstrip())
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]


def dfs(x, y):
    visited[x][y] = True
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx and nx < m and 0 <= ny and ny < n:
            if field[nx][ny] == 1 and not visited[nx][ny]:
                dfs(nx, ny)


for _ in range(t):
    m, n, k = map(int, sys.stdin.readline().rstrip().split())
    field = [[0] * n for _ in range(m)]
    visited = [[False] * n for _ in range(m)]

    for _ in range(k):
        x, y = map(int, sys.stdin.readline().rstrip().split())
        field[x][y] = 1

    cnt = 0

    for x in range(m):
        for y in range(n):
            if field[x][y] == 1 and not visited[x][y]:
                dfs(x, y)
                cnt += 1

    print(cnt)
